replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -14,147 +22,79 @@ private[journal] object Batch {
 
   def of(records: NonEmptyList[ActionRecord[Action]]): List[Batch] = {
 
-    // returns `true` when we can optimize Cassandra usage by NOT inserting "previous" Append actions in DB
-    // because "current" Delete action will discard them all
-    def cut(appends: Appends, delete: Action.Delete): Boolean = {
-      val append = appends.records.head.action
-      append.range.to <= delete.to.value
-    }
-
-    def dropDeleted(appends: Appends, delete: Action.Delete): Option[Appends] = {
-      // we cannot drop individual events in record as we do not parse the payload here (`records.head.action.payload: ByteVector`)
-      val validRecords = appends.records.collect { case record if record.action.range.to >= delete.to.value => record }
-      NonEmptyList.fromList(validRecords).map { Appends(appends.offset, _) }
-    }
+    val actions = records
+      .foldLeft(List.empty[ActionRecord[Action]]) {
+        // drop all `Mark` actions
+        case (acc, ActionRecord(_: Action.Mark, _)) => acc
 
-    records
-      .foldLeft(List.empty[Batch]) { (bs, record) =>
-        val offset = record.partitionOffset.offset
+        // drop all actions before `Purge`
+        case (_, r @ ActionRecord(_: Action.Purge, _)) => List(r)
 
-        def appendsOf(records: NonEmptyList[ActionRecord[Action.Append]]): Appends = {
-          Appends(offset, records)
-        }
-
-        def deleteOf(to: DeleteTo, origin: Option[Origin], version: Option[Version]): Delete = {
-          Delete(offset, to, origin, version)
-        }
-
-        def purgeOf(origin: Option[Origin], version: Option[Version]): Purge = {
-          Purge(offset, origin, version)
+        // collect `Append` and `Delete` actions
+        case (acc, record) => acc :+ record
```

Comment:
I do not know if this is important, but this append will take `O(n)` for a list of size `n`.

I.e. the whole filtering loop will take `O(n)`. I.e. if the batch contains `100` events then there will be `1000` appends.

As far as I remember these batches are much smaller, so this could be a premature optimization from my side, but might be worth to double check.

Reply:
yeah, I was thinking about benchmarking existing and new code, but didn't do that, yet.

in `ReplicateRecords` we do few other traversals over all records from `poll`, if I am reading code correctly, then:
* in `TopicReplicator:116` we get all `records` from Kafka's `poll` result
* on `TopicReplicator:120` we group them in to `SortedMap` them by `partition`
* on `TopicReplicator:145` we group them by `key`
* on `TopicReplicator:164` we feed all those per-key records to `ReplicateRecords`
* on `ReplicateRecords:172` we filter them
* on `ReplicateRecords:175` we pass them to `ReplicateRecords:161` where we pass them to `Batch` 

in sum, before `Batch`, we have traversed them all at least 3 times already, though in `Batch` we will traverse them 3 additional times 😞

Reply:
maybe it is not so bad - in current version:
* we process all records with `foldLeft`
* if there are `delete` actions, we backtrack
* then at `Batch:143` we reverse batched items

in new version:
* we do `foldLeft` and `groupBy`
* then we traverse 3 groups

looks like we will do one more traverse, though it is hard to compare, as new version is more aggresisve with batching of `delete` actions (which are rare)

performance-wise this change affects only `replicator` part - easy to allocate more resources, if required.

consistency-wise new code feels better.

comprehension-wise - at least for me - new code looks easier to read and understand, it doesn't have backtracking in few of 12 match-cases and there is no reverse-logic

Reply:
Doing more traverses is fine, because it just increments a constant, i.e. we keep it `O(n)`.

But having `:+` might lead to `O(n^2)`, which could be a problem for larger bets.

Reply:
checked on Live - according to metrics, which are scraped once a minute, on average across any minute we have batches with less than 1.2 records.

like it could be that we have 99 batches with size 1 and 1 batch with 20 records in it... over a minute we might have some really large batches, though realistically we rarely produce more than few records at time per aggregate.

Reply:
I am thinking about the theoretical situation, when there is only 1 record per batch, normally, and then, suddenly (because of the glitch on producer side), 10 000 records come in. Then we will end up with 100 000 000 cycles instead of 10 000 x C (where C is some constant).

Or the batch size is strictly limited, anyway?

Reply:
by default - max is 500 (default in Kafka's config [max-poll-records](https://docs.confluent.io/platform/current/installation/configuration/consumer-configs.html#max-poll-records)).

I could:
* run performance tests with this PR vs "no-op" `Batch`-ing (the one, which doesn't batch anything)
* rework the implementation to use prepend and reverse at the end
* reimplement code to use `Vector` or `ListBuffer` instead of `List` - those have better append performance

Reply:
so, `original` is faster, but it does less reductions on edge cases - leads to execution of more Cassandra statements

Reply:
the latest version:
```
[info] Benchmark                            Mode  Cnt       Score      Error  Units
[info] BatchBenchmark.alternativeReversed  thrpt    5  117159.312 ±  918.479  ops/s
[info] BatchBenchmark.original             thrpt    5  126303.502 ± 3433.752  ops/s
```

@rtar, what do you think about new approach with internal collection state? performance wise it is the same, and all unit-tests pass too